packages/store/src/internal/error-handler.ts

```diff
@@ -0,0 +1,69 @@
+import { ErrorHandler, Injectable, Injector } from '@angular/core';
+import { Observable } from 'rxjs';
+
+import { leaveNgxs } from '../operators/leave-ngxs';
+import { NgxsExecutionStrategy } from '../execution/symbols';
+
+/**
+ * This operator is used for piping the observable result
+ * from the `dispatch()`. It has a "smart" error handling
+ * strategy that allows us to decide whether we propagate
+ * errors to Angular's `ErrorHandler` or enable users to
+ * handle them manually. We consider following cases:
+ * 1) `store.dispatch()` (no subscribe) -> call `handleError()`
+ * 2) `store.dispatch().subscribe()` (no error callback) -> call `handleError()`
+ * 3) `store.dispatch().subscribe({ error: ... })` -> don't call `handleError()`
+ * 4) `toPromise()` without `catch` -> do `handleError()`
+ * 5) `toPromise()` with `catch` -> don't `handleError()`
+ */
+export function ngxsErrorHandler<T>(
+  internalErrorReporter: InternalErrorReporter,
+  ngxsExecutionStrategy: NgxsExecutionStrategy
+) {
+  return (source: Observable<T>) => {
+    let subscribed = false;
+
+    source.subscribe({
+      error: error => {
+        // Do not trigger change detection for a microtask. This depends on the execution
+        // strategy being used, but the default `DispatchOutsideZoneNgxsExecutionStrategy`
+        // leaves the Angular zone.
+        ngxsExecutionStrategy.enter(() =>
+          Promise.resolve().then(() => {
+            if (!subscribed) {
+              ngxsExecutionStrategy.leave(() =>
+                internalErrorReporter.reportErrorSafely(error)
+              );
+            }
+          })
+        );
+      }
+    });
+
+    return new Observable(subscriber => {
+      subscribed = true;
+      return source.pipe(leaveNgxs(ngxsExecutionStrategy)).subscribe(subscriber);
+    });
```

Comment:
This is a bit different to what we came up with when we paired on this.
Won't this result in the scenario where, if you `subscribe` to the dispatch but don't handle the error, then the error would not be handled globally (because the error handler on line 27 would prevent global propagation and the guard on line 33 would prevent the error handler from being hit)?
We would need to unsubscribe to line 26 when a subscriber arrives so that we can allow for the normal rxjs strategy to be used when errors are not handled in a subscription.